src/liballoc/string.rs

```diff
@@ -1452,6 +1453,20 @@ impl String {
         self.vec.clear()
     }
 
+    /// Converts the `String` into an iterator of `char`s.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let s = String::from("easy as α β γ");
+    /// assert_eq!(s.into_chars().count(), 13);
+    /// ```
+    #[unstable(feature = "into_chars", reason = "new API", issue="0")]
+    pub fn into_chars(self) -> IntoChars {
+        let chars = unsafe { mem::transmute(self.chars()) };
```

Comment:
Lying about lifetimes is not nice and there were some soundness issues related to that in the past. For example this can very silently become a read of freed memory if `_s` is freed and drop for `chars` happens to do something with the data it references.

If we’re doing this, can we do this properly, for example, by adding a new `CommonChars` into `core` that’s generic over the iterator and making the original `Chars` a `type Chars = CommonChars<SliceIter>` and the string’s `type Chars = core::CommonChars<vec::IntoIter<u8>>` or something?

Reply:
> For example this can very silently become a read of freed memory if _s is freed and drop for chars happens to do something with the data it references.

Yeah, that's why I commented that `Chars` doesn't implement `Drop` 😜 

> adding a new `CommonChars` into core that’s generic over the iterator

Certainly! I'm very much in favor of removing the `unsafe` here.

Reply:
> making the original `Chars` a `type Chars =`

I've done this, but it exposes the `CommonChars` in the documentation; is that acceptable?

Reply:
I think `CommonChars` should be private, or at least doc-hidden and perma-unstable. I think it follows that `Chars` should remain a struct with private fields, even if that involves adding trival forwarding impls.

(But regardless I remain skeptical that `String::into_chars` should exist in the first place, see other comment.)

Reply:
Yep, I figured we'd want to delegate all the parts, I just wanted to make sure others agreed before I put in that legwork.